lib/srv/alpnproxy/proxy.go

```diff
@@ -254,7 +255,8 @@ func New(cfg ProxyConfig) (*Proxy, error) {
 func (p *Proxy) Serve(ctx context.Context) error {
 	ctx, cancel := context.WithCancel(ctx)
 	defer cancel()
-	p.cancel = cancel
+	p.setCancelFunc(cancel)
```

Comment:
Pre-existing: guard against multiple Serve invocations? The interaction between multiple Serves and Stop gets messy if we allow it to happen. Granted, it shouldn't happen, but since we are here we might as well do it.

```suggestion
	ctx, cancel := context.WithCancel(ctx)
	defer cancel()

	p.mu.Lock()
	if p.cancel != nil {
		p.mu.Unlock()
		return trace.BadParameter("Serve may not be called twice")
	}
	p.cancel = cancel
	p.mu.Unlock()
```

If you do take this it may be better to do away with the new `func setCancelFunc` and `func cancel`.

Reply:
Done